src/Grpc.AspNetCore.Web/Internal/Base64PipeReader.cs

```diff
@@ -0,0 +1,170 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Buffers;
+using System.Buffers.Text;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Grpc.AspNetCore.Web.Internal
+{
+    /// <summary>
+    /// Reads and decodes base64 encoded bytes from the inner reader.
+    /// </summary>
+    internal class Base64PipeReader : PipeReader
+    {
+        private readonly PipeReader _inner;
+        private ReadOnlySequence<byte> _currentInnerBuffer;
+        private ReadOnlySequence<byte> _currentDecodedBuffer;
+
+        public Base64PipeReader(PipeReader inner)
+        {
+            _inner = inner;
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            var consumedPosition = ResolvePosition(consumed);
+
+            _inner.AdvanceTo(consumedPosition);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            var consumedPosition = ResolvePosition(consumed);
+            var examinedPosition = ResolvePosition(examined);
+
+            _inner.AdvanceTo(consumedPosition, examinedPosition);
+        }
+
+        private SequencePosition ResolvePosition(SequencePosition base64Position)
+        {
+            // TODO - is there a better way to get the index of a position in a sequence?
+            var length = _currentDecodedBuffer.Slice(0, base64Position).Length;
+
+            var resolvedPosition = _currentInnerBuffer.GetPosition(Base64.GetMaxEncodedToUtf8Length((int)length));
+            return resolvedPosition;
+        }
+
+        public override void CancelPendingRead()
+        {
+            _inner.CancelPendingRead();
+        }
+
+        public override void Complete(Exception? exception = null)
+        {
+            _inner.Complete(exception);
+        }
+
+        public async override ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            var innerResult = await _inner.ReadAsync(cancellationToken);
+            if (innerResult.Buffer.IsEmpty)
+            {
+                _currentDecodedBuffer = innerResult.Buffer;
+                _currentInnerBuffer = innerResult.Buffer;
+                return innerResult;
+            }
+
+            // Minimum valid base64 length is 4. Read until we have at least that much content
+            while (innerResult.Buffer.Length < 4)
```

Comment:
For CancelPendingRead to work, we should check every ReadResult and check if it's canceled and return to the caller if it is. Otherwise, the developer using the Base64PipeReader might call CancelPendingRead, but never see a canceled read result. In the case we receive a canceled ReadResult before reading the minimum about of encoded data, I would make sure to return an empty-but-still-canceled ReadResult.

Reply:
Done